File: lib/aura/kernel/event_bus.rb
# frozen_string_literal: true

module Aura
  module Kernel
    # A generic publisher-subscriber event bus.
    class EventBus
      def initialize
        @listeners = Hash.new { |h, k| h[k] = [] }
      end

      def subscribe(event_type, &block)
        @listeners[event_type] << block
        self
      end

      def emit(event_type, **payload)
        @listeners[event_type].each do |listener|
          begin
            listener.call(payload)
          rescue => e
            $stderr.puts "[EventBus Error] Exception in listener for #{event_type}: #{e.message}"
          end
        end
        # Also support wildcard listeners
        @listeners[:*].each do |listener|
          begin
            listener.call(event_type, payload)
          rescue => e
            $stderr.puts "[EventBus Error] Exception in wildcard listener for #{event_type}: #{e.message}"
          end
        end
      end
    end

    # Adapts traditional Bridge block callback hashes to the EventBus architecture.
    class CallbackEventBus
      def initialize(callbacks = {})
        @callbacks = callbacks || {}
      end

      def emit(event, **payload)
        case event
        when :plan_event
          handle_plan_event(payload)
        when :final_answer
          @callbacks[:on_final_answer]&.call(payload[:content])
        when :tool_halted
          @callbacks[:on_warning]&.call("Tool '#{payload[:tool]}' halted (#{payload[:status]}): #{payload[:advice]}")
        when :loop_aborted
          @callbacks[:on_warning]&.call("Agent loop aborted: #{payload[:reason]}")
        end
      end

      private

      def handle_plan_event(payload)
        case payload[:type]
        when "delta"
          @callbacks[:on_token]&.call(payload[:text].to_s)
        end
      end
    end

    class NullEventBus
      def emit(event, **); end
    end
  end
end

File: src/core/memory/eventBus.ts
import { AgentEventBus, type EventEmitterLike } from '../kernel/eventBus.js';

export type { EventEmitterLike };

/**
 * MemoryEventBus inherits from AgentEventBus for backward compatibility.
 */
export class MemoryEventBus extends AgentEventBus {
  constructor(emitter?: EventEmitterLike) {
    super(emitter);
  }
}

// Alias for Minitest/Ruby compatibility
export const EventBus = MemoryEventBus;

interface Callbacks {
  on_token?: (text: string) => void;
  on_final_answer?: (content: string) => void;
  on_warning?: (message: string) => void;
}

export class CallbackEventBus {
  private callbacks: Callbacks;

  constructor(callbacks: Callbacks = {}) {
    this.callbacks = callbacks || {};
  }

  public emit(event: string, payload: Record<string, unknown> = {}): void {
    if (!this.callbacks) return;

    switch (event) {
      case 'plan_event':
        if (
          payload?.type === 'delta' &&
          typeof this.callbacks.on_token === 'function'
        ) {
          this.callbacks.on_token(
            payload.text !== undefined ? String(payload.text) : '',
          );
        }
        break;
      case 'final_answer':
        if (typeof this.callbacks.on_final_answer === 'function') {
          this.callbacks.on_final_answer(payload?.content as string);
        }
        break;
      case 'tool_halted':
        if (typeof this.callbacks.on_warning === 'function') {
          this.callbacks.on_warning(
            `Tool '${payload?.tool}' halted (${payload?.status}): ${payload?.advice || ''}`,
          );
        }
        break;
      case 'loop_aborted':
        if (typeof this.callbacks.on_warning === 'function') {
          this.callbacks.on_warning(
            `Agent loop aborted: ${payload?.reason || ''}`,
          );
        }
        break;
    }
  }
}

export class NullEventBus {
  public emit(_event: string, _payload?: Record<string, unknown>): void {}
}
